shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,317 @@
+     * basis. These tasks will persist until the provider that they use emits a done task, or the task is re-emitted
+     * with a priority of 0.
+     *
+     * Subtasks:
+     * These are created when a provider task emits a task to complete. These tasks must be emitted each time that
+     * reaction is run to persist as if that reaction runs again and does not emit these tasks they will be
+     * cancelled. Within these tasks the priority is used to break ties between two subtasks that share the same
+     * root task. In these cases the subtask that requested the provider with the highest priority will have its
+     * task executed. The other subtask will be blocked until that task is no longer in the call queue.
+     *
+     * If a subtask is emitted with optional then it is compared differently to other tasks when it comes to priority.
+     * This task and all its decendents will be considered optional. If it is compared to a task that does not have
+     * optional in it's parantage, the non optional task will win. However decendents of this task that are not optional
+     * will compare to eachother as normal.
+     *
+     * @tparam T the provider type that this task is for
+     */
+    template <typename T>
+    struct Task {
+
+        static void emit(NUClear::PowerPlant& powerplant,
+                         std::shared_ptr<T> data,
+                         const std::string& name = "",
+                         int priority            = 1,
+                         bool optional           = false) {
+
+            // Work out who is sending the task so we can determine if it's a subtask
+            auto* task           = NUClear::threading::ReactionTask::get_current_task();
+            uint64_t reaction_id = task ? task->parent.id : -1;
+            uint64_t task_id     = task ? task->id : -1;
+
+            NUClear::dsl::word::emit::Direct<T>::emit(powerplant,
+                                                      std::make_shared<commands::DirectorTask>(typeid(T),
+                                                                                               reaction_id,
+                                                                                               task_id,
+                                                                                               data,
+                                                                                               name,
+                                                                                               priority,
+                                                                                               optional));
+        }
+    };
+
+    /**
+     * This is a special task that should be emitted when a provider finishes the task it was given.
+     * When this is emitted the director will re-execute the provider which caused this task to run.
+     *
+     * ```
+     * emit<Task>(std::make_unique<Done>());
+     * ```
+     */
+    struct Done {};
```

Comment:
Another case where I am curious why you have elected against a protobuf message

Reply:
I prefer extensions to be self-contained and not include messages.
That way all the code for this extension is together.

Like how Configuration isn't done with a proto message even though it maybe could have been.